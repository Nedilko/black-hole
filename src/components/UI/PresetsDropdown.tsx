import { memo } from 'react';
import { MdKeyboardArrowDown } from 'react-icons/md';
import { DIFFUCULTY } from '../../store/gameSlice';
import { getPresetName } from '../../store/presets';

type PropsType = {
  value: DIFFUCULTY;
  handleChange: (value: DIFFUCULTY) => void;
};

const PresetsDropdown = ({ value, handleChange }: PropsType) => {
  const changeValueHandler = (e: React.ChangeEvent<HTMLSelectElement>) => {
    handleChange(+e.target.value as DIFFUCULTY);
  };

  return (
    <label className="flex text-gray-300 uppercase relative w-full">
      <MdKeyboardArrowDown className="absolute translate-y-1/2 right-0 cursor-pointer text-3xl pointer-events-none" />
      <div className="text-sm absolute top-1 left-2 text-gray-400">
        difficulty
      </div>
      <select
        name="presets"
        id="presets"
        className="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-normal bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150 hover:cursor-pointer"
        onChange={changeValueHandler}
        value={value}
      >
        <option value={DIFFUCULTY.EASY}>
          {getPresetName(DIFFUCULTY.EASY)}
        </option>
        <option value={DIFFUCULTY.MEDIUM}>
          {getPresetName(DIFFUCULTY.MEDIUM)}
        </option>
        <option value={DIFFUCULTY.HARD}>
          {getPresetName(DIFFUCULTY.HARD)}
        </option>
        <option value={DIFFUCULTY.CUSTOM}>
          {getPresetName(DIFFUCULTY.CUSTOM)}
        </option>
      </select>
    </label>
  );
};

export default memo(PresetsDropdown);
