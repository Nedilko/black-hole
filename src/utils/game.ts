import type { Results } from '../store/gameSlice';
export enum DIFFUCULTY {
  EASY,
  MEDIUM,
  HARD,
  CUSTOM,
}

export const getDefaultResults = (): Results => {
  return {
    [DIFFUCULTY.EASY]: 0,
    [DIFFUCULTY.MEDIUM]: 0,
    [DIFFUCULTY.HARD]: 0,
    [DIFFUCULTY.CUSTOM]: 0,
  };
};
