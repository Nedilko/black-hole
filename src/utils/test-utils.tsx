import { RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from 'redux';
import { AppStore, RootState, setupStore } from '../store';
import { render } from '@testing-library/react';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = setupStore(initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
