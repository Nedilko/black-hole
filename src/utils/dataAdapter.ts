import type { Results } from '../store/gameSlice';
import { getDefaultResults } from './game';
import {
  writeToLocalstorage,
  readFromLocalstorage,
} from './localstorageAdapter';

const loadResults = (): Results => {
  const localStorageData = readFromLocalstorage('results');
  if (!localStorageData) {
    writeResults(getDefaultResults());
    return getDefaultResults();
  }
  return JSON.parse(atob(localStorageData));
};

const writeResults = (settings: Results): void => {
  writeToLocalstorage('results', btoa(JSON.stringify(settings)));
};

export { loadResults, writeResults };
