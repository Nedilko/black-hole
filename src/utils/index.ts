import { getTime } from './time';
import {
  readFromLocalstorage,
  writeToLocalstorage,
} from './localstorageAdapter';
import { getDefaultResults } from './game';
import { loadResults, writeResults } from './dataAdapter';

export {
  getTime,
  readFromLocalstorage,
  writeToLocalstorage,
  getDefaultResults,
  loadResults,
  writeResults,
};
