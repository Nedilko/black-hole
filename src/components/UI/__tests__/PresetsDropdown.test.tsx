import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import PresetsDropdown from '../PresetsDropdown';
import userEvent from '@testing-library/user-event';

describe('Header', () => {
  const handler = jest.fn();

  beforeEach(() => {
    handler.mockClear();
  });

  it('should render PresetsDropdown', () => {
    const { container } = renderWithProviders(
      <PresetsDropdown value={0} handleChange={handler} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render with selected value', () => {
    renderWithProviders(<PresetsDropdown value={1} handleChange={handler} />);
    expect(
      (screen.getByTestId('presets-dropdown') as HTMLSelectElement).value
    ).toBe('1');
  });

  it('should call handler on change', () => {
    renderWithProviders(<PresetsDropdown value={0} handleChange={handler} />);

    userEvent.selectOptions(screen.getByTestId('presets-dropdown'), '2');

    expect(handler).toHaveBeenCalledTimes(1);
    expect(handler).toHaveBeenCalledWith(2);
  });
});
