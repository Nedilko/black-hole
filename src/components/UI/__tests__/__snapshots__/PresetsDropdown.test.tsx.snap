// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render PresetsDropdown 1`] = `
<div>
  <label
    class="flex text-gray-300 uppercase relative w-full"
  >
    <svg
      class="absolute translate-y-1/2 right-0 cursor-pointer text-3xl pointer-events-none"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
      />
    </svg>
    <div
      class="text-sm absolute top-1 left-2 text-gray-400"
    >
      difficulty
    </div>
    <select
      class="appearance-none outline outline-1 hover:outline-2 outline-slate-700 focus:outline-gray-400 outline-offset-2 w-full text-xl font-normal bg-slate-900 px-2 pt-6 pb-1 rounded transition-all duration-150 hover:cursor-pointer"
      data-testid="presets-dropdown"
      name="presets-dropdown"
    >
      <option
        value="0"
      >
        Easy
      </option>
      <option
        value="1"
      >
        Medium
      </option>
      <option
        value="2"
      >
        Hard
      </option>
      <option
        value="3"
      >
        Custom
      </option>
    </select>
  </label>
</div>
`;
